ReactNativeClient/lib/services/searchengine/queryBuilder.ts

```diff
@@ -0,0 +1,505 @@
+		)`);
+		params.push(locationTerm.value);
+	});
+};
+
+const addExcludeTextConditions = (excludedTerms: Term[], conditions:string[], params: string[], relation: string) => {
+	const type = excludedTerms[0].name;
+
+	if (excludedTerms && relation === 'AND') {
+		conditions.push(`
+		AND ROWID NOT IN (
+			SELECT ROWID
+			FROM notes_fts
+			WHERE notes_fts.${type} MATCH ?
+		)`);
+		params.push(excludedTerms.map(x => x.value).join(' OR '));
+	}
+
+	if (excludedTerms && relation === 'OR') {
+		excludedTerms.forEach(term => {
+			conditions.push(`
+			OR ROWID IN (
+				SELECT *
+				FROM (
+					SELECT ROWID
+					FROM notes_fts
+					EXCEPT
+					SELECT ROWID
+					FROM notes_fts
+					WHERE notes_fts.${type} MATCH ?
+				)
+			)`);
+			params.push(term.value);
+		});
+	}
+};
+
+
+const textFilter = (filters: Term[], conditions: string[], params: string[], relation: string) => {
+	const allTerms = filters.filter(x => x.name === 'title' || x.name === 'body' || x.name === 'text');
+
+	const includedTerms = allTerms.filter(x => !x.negated);
+	if (includedTerms.length > 0) {
+		conditions.push(`${relation} notes_fts MATCH ?`);
+		const termsToMatch = includedTerms.map(term => {
+			if (term.name === 'text') return term.value;
+			else return `${term.name}:${term.value}`;
+		});
+		const matchQuery = (relation === 'OR') ? termsToMatch.join(' OR ') : termsToMatch.join(' ');
+		params.push(matchQuery);
+	}
+
+	const excludedTextTerms = allTerms.filter(x => x.name === 'text' && x.negated);
+	const excludedTitleTerms = allTerms.filter(x => x.name === 'title' && x.negated);
+	const excludedBodyTerms = allTerms.filter(x => x.name === 'body' && x.negated);
+
+	if ((excludedTextTerms.length > 0) && relation === 'AND') {
+		conditions.push(`
+		AND ROWID NOT IN (
+		SELECT ROWID
+		FROM notes_fts
+		WHERE notes_fts MATCH ?
+		)`);
+		params.push(excludedTextTerms.map(x => x.value).join(' OR '));
+	}
+
+	if ((excludedTextTerms.length > 0) && relation === 'OR') {
+		excludedTextTerms.map(textTerm => {
+			conditions.push(`
+			OR ROWID IN (
+			SELECT *
+			FROM (
+				SELECT ROWID
+				FROM notes_fts
+				EXCEPT
+				SELECT ROWID FROM notes_fts
+				WHERE notes_fts MATCH ?
+			))`);
+			params.push(textTerm.value);
+		});
+	}
+
+	if (excludedTitleTerms.length > 0) {
+		addExcludeTextConditions(excludedTitleTerms, conditions, params, relation);
+	}
+
+	if (excludedBodyTerms.length > 0) {
+		addExcludeTextConditions(excludedBodyTerms, conditions, params, relation);
+	}
+
+};
+
+const getDefaultRelation = (filters: Term[]): 'OR' | 'AND' => {
+	const anyTerm = filters.find(term => term.name === 'any');
+	if (anyTerm) { return (anyTerm.value === '1') ? 'OR' : 'AND'; }
+	return 'AND';
+};
+
+const getConnective = (filters: Term[], relation: 'OR' | 'AND'): string => {
+	const notebookTerm = filters.find(x => x.name === 'notebook');
+	return (!notebookTerm && (relation === 'OR')) ? 'ROWID=-1' : '1'; // ROWID=-1 acts as 0 (something always false)
+};
+
```

Comment:
I would suggest making the relation an enum as that way you won't need to repeat AND | OR a bit everywhere. And the code is not consistent because here it's "AND | OR" but other places it's just "string". If you define an enum, it makes it much clearer what are the possible values, and it can prevent errors.